src/linear_algebra/unitary_group.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Shing Tak Lam
+-/
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import data.complex.basic
+
+/-!
+# The Unitrary Group O(n)
+
+This file defines elements of the unitary group `unitary_group n`, also written `U(n)`,
+consisting of all `n` by `n` complex matrices where the conjugate transpose is the inverse.
+In addition, we define the group structure on `unitary_group n` and the embedding into the
+general linear group `general_linear_group ℂ (n → ℂ)`.
+
+## Main Definitions
+
+ * `matrix.unitary_group` is the type of matrices where the transpose is the inverse
+ * `matrix.unitary_group.group` is the group structure (under multiplication)
+ * `matrix.unitary_group.embedding_GL` is the embedding `U(n) → GLₙ(ℝ)`
+
+## References
+
+ * https://en.wikipedia.org/wiki/Unitary_group
+
+## Tags
+
+matrix group, group, unitary group
+
+-/
+
+namespace matrix
+universe u
+open linear_map
+
+section
+
+variables {n m : Type*} [fintype n] [fintype m] [decidable_eq n] [decidable_eq m]
+
+/--
+  The conjugate transpose of a matrix, defined to be a the element-wise conjugate of the transpose.
+-/
+def conj_transpose (M : matrix n m ℂ) : matrix m n ℂ :=
+  M.transpose.map complex.conj
+
+localized "postfix `†`:1500 := conj_transpose" in matrix
+
+lemma conj_transpose_eq_transpose_map_conj (M : matrix n m ℂ) :
+  M† = (M.map complex.conj).transpose :=
+by ext; simp only [conj_transpose, map_apply, transpose_apply]
+
+@[simp]
+lemma conj_transpose_mul {p : Type*} [fintype p] [decidable_eq p] (M : matrix n m ℂ)
+  (N : matrix m p ℂ) : (matrix.mul M N)† = matrix.mul N† M† := by simp [conj_transpose]
+
+@[simp]
+lemma conj_transpose_one : (1 : matrix n n ℂ)† = 1 := by simp [conj_transpose]
+
+@[simp]
+lemma conj_transpose_conj_transpose (M : matrix n m ℂ) : M†† = M :=
+by simp [conj_transpose, transpose_map, function.involutive.comp_self complex.conj_involutive]
+
+end
+
+open_locale matrix
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n]
+
+/--
+  `unitary_group n` is the group of `n` by `n` complex matrices where the conjugate transpose is
+  the inverse.
+-/
+def unitary_group := {M : matrix n n ℂ // matrix.mul M† M = 1}
```

Comment:
Here `\bub` gives the wrong instance? If I change it to `\bub` then a few of the lemmas below break.

Reply:
Nevermind, it's not `bub`